Add option to use Symfony Dependency Injection as Container

marcosh opened the pull request:
Symfony Dependency Injection now implements PSR-11 so it is compatible with Expressive.

I provide here a working configuration to use as container

Ocramius commented:
Awesome!

stof commented on src/ExpressiveInstaller/Resources/config/container-symfony-di.php:
```diff
@@ -0,0 +1,102 @@
+
+// Inject invokables
+foreach ($config['dependencies']['invokables'] as $name => $object) {
+    $container->register($name, new $object);
```
This is wrong. The second argument of `register` is the class name

stof commented on src/ExpressiveInstaller/Resources/src/CallableFactory.php:
```diff
@@ -0,0 +1,53 @@
+
+namespace App;
+
+use Interop\Container\ContainerInterface;
```
why not using PSR-11 instead ?

marcosh replied:
the test (see https://github.com/zendframework/zend-expressive-skeleton/blob/master/test/ExpressiveInstallerTest/ContainersTest.php#L80) at the moment requires explicitely `Container\Interop\ContainerInterop`.

Another option would be to change the test and ask that `$container` is instance either of `Container\Interop\ContainerInterop` or `Psr\Container\ContainerInterface`

stof replied:
I'm commenting on the use statement for the typehint in the factory, not for the container being returned.

And PSR-11 ContainerInterface is a parent interface of the Interop one.

xtreamwayz replied:
There is #160 to change to `Psr\Container\ContainerInterface`. As soon as I have some spare time I'll merge it.

xtreamwayz commented:
@marcosh I've merged some PR's. Could you please rebase onto develop?

marcosh commented:
@xtreamwayz sure, as soon as find some time I will rebase and adjust some things

stof commented:
Note that this setup is not the recommended way to use Symfony DI if you care about performance. You are using the ContainerBuilder at runtime, which means you compile the container for each request.
Symfony DI is optimized for building the container once, and then dumping an optimized generated container in some cache, to use it at runtime. As such, the compilation phase of the container is costly (and becomes more costly with each new feature we add, as we try to push as much work as possible to the compilation phase rather than the runtime).

Anyone caring about performance should not use the container exposed in this PR.

xtreamwayz commented:
@stof Thanx for the explanation.

So I guess there must be a caching mechanism in place before this is usable in production.

marcosh commented:
@xtreamwayz I rebased and fixed some things here and there

@stof do you recommend to implement something like what is described [here](https://symfony.com/doc/current/components/dependency_injection/compilation.html#dumping-the-configuration-for-performance)? Or maybe just document the issue and provide a link to the documentation?

xtreamwayz commented:
Don't link to the issue. Implement it. :) 
This is a skeleton where people expect things to be ready out of the box.

marcosh commented:
great! will do!

stof commented:
Note however that the usage of closures does not play well with dumping the container (closures are very hard to dump when doing code generation, and Symfony does not support doing it).
I don't know enough about the format of the config of expressive to know whether using a dumped container is doable.

stof commented:
btw, I'm not sure maintaining the whole logic of converting an Expressive config to the Symfony DIC in the skeleton is the best idea: people would have to synchronize the logic manually in their project to get get any bugfix being done. It might be worth extracting the logic to a separate package, which can be installed as a dependency (and so getting bugfixes is a matter of `composer update`) and having the skeleton using this package instead, so that the logic living in the skeleton is small.
Skeletons are bad at providing lots of logic, because of the update process of this logic.

xtreamwayz commented:
The expressive config itself is dumped. But that only works if you don't use closures. If you stick to factories it isn't a problem.

However, I don't know how this would work with Symfony DI since I've never used it as a standalone package.

marcosh commented:
@xtreamwayz @stof  I implemented a caching mechanism and it seems to work

stof commented on src/ExpressiveInstaller/Resources/src/ExpressiveSymfonyContainer.php:
```diff
@@ -0,0 +1,226 @@
+
+        $this->cacheContainer($container);
+
+        return $container;
```
you should not return the container builder. You should also return the optimized container you just generated in the cache

stof commented on src/ExpressiveInstaller/Resources/src/ExpressiveSymfonyContainer.php:
```diff
@@ -0,0 +1,226 @@
+
+        // Register the container in the container
+        // In this way in can be referenced in factories
+        $container->register('container')->setSynthetic(true);
```
this does not make sense. Symfony already exposes the container as a service (named `service_container`), and optimizes the dumping. If you want to have it available as `container`too, create an alias instead

stof commented on src/ExpressiveInstaller/Resources/src/ExpressiveSymfonyContainer.php:
```diff
@@ -0,0 +1,226 @@
+
+        // Inject config
+        $container->register('config')->setSynthetic(true);
+        $container->set('config', $this->config);
```
Values of synthetic services must be set when *using* the container, not when building it. Otherwise, they are missing when using the cache.

stof commented on src/ExpressiveInstaller/Resources/src/ExpressiveSymfonyContainer.php:
```diff
@@ -0,0 +1,225 @@
+            return;
+        }
+
+        if (! (isset($this->config[static::ENABLE_CACHE]) && $this->config[static::ENABLE_CACHE])) {
```
you should always dump the container in the cache IMO. However, you should support the debug mode of Symfony (look at how the Kernel class implements it), which automatically refreshes the cache when needed in development

stof commented on src/ExpressiveInstaller/Resources/src/ExpressiveSymfonyContainer.php:
```diff
@@ -0,0 +1,225 @@
+        foreach ($this->config['dependencies']['factories'] as $name => $object) {
+            $container->register($name)
+                ->setClass($object)
+                ->addArgument(new Reference('container'))
```
use the builtin `service_container` id.

stof commented on src/ExpressiveInstaller/Resources/src/ExpressiveSymfonyContainer.php:
```diff
@@ -0,0 +1,225 @@
+    }
+
+    /**
+     * return Container
```
broken phpdoc

marcosh commented:
@stof I think I fixed the things you suggested.

I left out, at least for the moment, the debug mode you mentioned

stof commented:
@marcosh leaving the debug mode out means you break the experience when developing (asking people to delete the cache file any time they change the config is not good, and always deleting the cache to rebuild the same is wasting their time)

marcosh commented:
@stof there is no need to delete by hand the file. It's enough to set the configuration key `ExpressiveSymfonyContainer::ENABLE_CACHE` to false and the container will be generated again every time.

You are right that this could be improved checking if the container needs to be renewed before actually regenerating it

stof commented:
@marcosh disabling cache is precisely the second case I mentionned: wasting developer time making them wait on each request, why the Symfony DI component has all features needed to avoid this (Symfony HttpKernel does not have an option to disable the usage of a cached container btw, and this is working fine for us since 7 years).
The only thing you might need is implementing a special ResourceInterface in case none of the existing ones allow you to detect changes in the expressive config (as you process the config at built time, changing it requires invalidating)

stof commented on src/ExpressiveInstaller/Resources/src/ExpressiveSymfonyContainer.php:
```diff
@@ -0,0 +1,192 @@
+        $container = $this->loadFromCache($this->cachedContainerFile);
+
+        // Inject config
+        $container->set('config', $this->config);
```
Should the config array really be set as a service (btw, arrays are not really meant to be added as services in Symfony, but as parameters).
And does `config['dependencies']` really need to be set here ? It is processed to build the container.

Btw, you could even set the config at build time instead of injecting it at runtime (unless you invalidate the cache only when `config['dependencies']` changes rather than the whole config. But I don't know what else you could have in this config)

xtreamwayz replied:
The `config` needs to be available in the container. It doesn't matter how it's done as long as this works: `$container->get('config');`

stof replied:
but does it need to be able to access `$config['dependencies']` ? If yes, you have an issue: in the request building the cache, you mutate the config, and so you inject a different one

xtreamwayz replied:
No, the `dependencies` key is only for configuring the container.

This is for zend-servicemanager.
```php
// Load configuration
$config = require __DIR__ . '/config.php';

// Build container
$container = new ServiceManager();
(new Config($config['dependencies']))->configureServiceManager($container);

// Inject config
$container->setService('config', $config);
```
It loads the config, grabs the dependencies to build the container and then injects the config so factories can access it. I can't think of any reason why someone would need access to the `dependencies` key.

Besides the dependencies the complete config for the application is in that `$config` array. An example of a [template config](https://github.com/xtreamwayz/xtreamwayz.com/blob/master/config/autoload/templates.global.php).

Some other containers don't accept the `$config` as an array and they convert it to an ArrayObject:
```php
new ArrayObject($this->config, ArrayObject::ARRAY_AS_PROPS)```

stof commented on src/ExpressiveInstaller/Resources/src/ExpressiveSymfonyContainer.php:
```diff
@@ -0,0 +1,192 @@
+            if (isset($this->config['dependencies']['services'][$service])) {
+                // Marshal from service
+                $instance = $this->config['dependencies']['services'][$service];
+                $factory = function () use ($instance) {
```
using a closure is not compatible with dumping.

Instead of using your ExpressiveSymfonyDelegatorFactory passing it a closure, your should let the service be registered normally (so remove the code here which unsets the config), and use Symfony's service decoration feature to apply delegators.

xtreamwayz commented:
I'm closing this PR. We experienced some bugs in the code for the Aura.DI container and we can't push fixes as that code isn't in a package. Therefor we are moving the code for Aura.DI and Pimple into their own packages in #177. We would like to see the same approach for this container.

Once the package is created, tested and stable, you can create a new PR or we can help you out integrating it into the installer.
